presto-hive/src/main/java/com/facebook/presto/hive/S3SelectLineRecordReader.java

```diff
@@ -0,0 +1,222 @@
+
+import static com.facebook.presto.hive.RetryDriver.retry;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_MAX_BACKOFF_TIME;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_MAX_CLIENT_RETRIES;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_MAX_RETRY_TIME;
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.apache.hadoop.hive.serde.serdeConstants.FIELD_DELIM;
+import static org.apache.hadoop.hive.serde.serdeConstants.LINE_DELIM;
+import static org.apache.hadoop.hive.serde.serdeConstants.SERIALIZATION_FORMAT;
+import static org.apache.http.HttpStatus.SC_BAD_REQUEST;
+import static org.apache.http.HttpStatus.SC_FORBIDDEN;
+import static org.apache.http.HttpStatus.SC_NOT_FOUND;
+
+@ThreadSafe
+public abstract class S3SelectLineRecordReader
+        implements RecordReader<LongWritable, Text>
+{
+    private InputStream selectObjectContent;
+    private long processedRecords;
+    private long recordsFromS3;
+    private long position;
+    private LineReader reader;
+    private boolean isFirstLine;
+    private static final Duration BACKOFF_MIN_SLEEP = new Duration(1, SECONDS);
+    private final PrestoS3SelectClient selectClient;
+    private final long start;
+    private final long end;
+    private final int maxAttempts;
+    private final Duration maxBackoffTime;
+    private final Duration maxRetryTime;
+    private final Closer closer = Closer.create();
+    private final SelectObjectContentRequest selectObjectContentRequest;
+    final String lineDelimiter;
+
+    S3SelectLineRecordReader(
+            Configuration configuration,
+            HiveClientConfig clientConfig,
+            Path path,
+            long start,
+            long length,
+            Properties schema,
+            String ionSqlQuery,
+            PrestoS3ClientFactory s3ClientFactory)
+    {
+        requireNonNull(configuration, "configuration is null");
+        requireNonNull(clientConfig, "clientConfig is null");
+        requireNonNull(schema, "schema is null");
+        requireNonNull(path, "path is null");
+        requireNonNull(ionSqlQuery, "ionSqlQuery is null");
+        requireNonNull(s3ClientFactory, "s3ClientFactory is null");
+        this.selectClient = new PrestoS3SelectClient(configuration, clientConfig, s3ClientFactory);
+        closer.register(selectClient);
+        this.lineDelimiter = (schema).getProperty(LINE_DELIM, "\n");
+        this.processedRecords = 0;
+        this.recordsFromS3 = 0;
+        this.start = start;
+        this.position = this.start;
+        this.end = this.start + length;
+        this.isFirstLine = true;
+        this.selectObjectContentRequest = buildSelectObjectRequest(schema, ionSqlQuery, path);
```

Comment:
Yes it could throw:
```
throw new PrestoException(NOT_SUPPORTED, "Compression extension not supported for S3 Select: " + path);
```

Moved client construction to the end.